datafusion/common/src/error.rs

```diff
@@ -83,6 +83,30 @@ pub enum DataFusionError {
     #[cfg(feature = "jit")]
     /// Error occurs during code generation
     JITError(ModuleError),
+    /// Error with context
```

Comment:
```suggestion
    /// Error with additional context
```

I am +0 on this change (as in I don't oppose it but I also would be fine without it) 

I can see the usecase (to keep the structure of nested errors), however there are other more structured approaches to Error handling (e.g https://crates.io/crates/snafu or https://crates.io/crates/thiserror), but so far DataFusion just uses string handling (aka `format!("Error processing subqueries: {}", inner_error)`);

Other libraries have quite sophisticated  `Error` hierarchies which theoretically allow the errors to be programmatically handled in different ways.  It adds non trivial maintenance and contribution overhead,  and I haven't seen anyone ask for this yet in DataFusion so I don't think it is worth adding yet.

Reply:
I would +1 any solution that allows nested errors (our own, or off-the-shelf). This was inspired by heavily commented code like:

```
    let subqry_input = only_or_err(subqry_inputs.as_slice()); // correlated where in subqueries must project a single expression
```

Turning into something that also benefits the user when it breaks like:

```
    let subqry_input = only_or_err(subqry_inputs.as_slice())
        .map_err(|e| context!("single expression projection required", e))?;
```

With the `context!` macro providing the relevant information about which optimizer was being run, and turned the error from `more than one item found in list!` into `"single expression projection required in decorrelate_where_in line 123" caused by "more than one item found in list!"` which seemed like a much better user experience.